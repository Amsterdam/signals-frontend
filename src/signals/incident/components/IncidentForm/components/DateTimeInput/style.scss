.datetime-input {
  position: relative;
  margin-top: -20px;

  &__label {
    margin: 8px 0 3px;
  }

  &__earlier-date {
    width: 240px;
  }

  &__earlier-time {
    &-hours,
    &-minutes {
      width: 80px !important;
    }

    &-label {
      margin: 0 16px 0 8px;
    }
  }
}
