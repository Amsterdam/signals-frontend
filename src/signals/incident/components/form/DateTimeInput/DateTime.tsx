import { useCallback, useEffect, useState } from 'react'
import type { FC } from 'react'

import { Label, RadioGroup } from '@amsterdam/asc-ui'
import Radio from 'components/RadioButton'
import Select from 'components/Select'
import format from 'date-fns/format'
import locale from 'date-fns/locale/nl'
import parse from 'date-fns/parse'
import subDays from 'date-fns/subDays'
import { capitalize } from 'shared/services/date-utils'
import type { Incident } from 'types/incident'

import {
  FieldWrapper,
  Info,
  StyledLabel,
  TimeFieldset,
  TimeWrapper,
} from './styled'

type DateIndicator = {
  id: 'now' | 'earlier'
  label: 'Nu' | 'Eerder'
}

const dateIndicators: DateIndicator[] = [
  {
    id: 'now',
    label: 'Nu',
  },
  {
    id: 'earlier',
    label: 'Eerder',
  },
]

export const defaultTimestamp = new Date()
defaultTimestamp.setHours(9)
defaultTimestamp.setMinutes(0)
defaultTimestamp.setSeconds(0)

const getFormattedDate = (date: Date) =>
  capitalize(format(date, 'EEEE d MMMM', { locale }))

type FormattedDate = 'Vandaag' | ReturnType<typeof getFormattedDate>

const formatDate = (
  offset: number,
  type: 'label' | 'value' = 'value'
): FormattedDate => {
  if (type === 'label' && offset === 0) {
    return 'Vandaag'
  }

  const date = subDays(new Date(), offset)

  return getFormattedDate(date)
}

export const daysOptions = [...Array(7).keys()].map((offset) => {
  const name = formatDate(offset, 'label')

  return {
    key: name,
    name,
    value: formatDate(offset),
  }
})

export const hoursOptions = [...Array(24).keys()].map((value) => ({
  key: value.toString(),
  name: value.toString(),
  value: value.toString(),
}))

export const minutesOptions = [...Array(12).keys()].map((minute) => ({
  key: (minute * 5).toString(),
  name: (minute * 5).toString(),
  value: (minute * 5).toString(),
}))

export interface DateTimeProps {
  onUpdate: (timestamp: Incident['timestamp']) => void
  value: Incident['timestamp']
}

type DateIndication = DateIndicator['id'] | ''

const dateIndicationValue: Record<string, DateIndication> = {
  undefined: '',
  object: 'now',
  number: 'earlier',
}

const DateTime: FC<DateTimeProps> = ({ onUpdate, value }) => {
  const [datetime, setDatetime] = useState(
    value ? new Date(value) : defaultTimestamp
  )

  const [dateIndication, setDateIndication] = useState<DateIndication>(
    getDateIndication(value)
  )

  function getDateIndication(value: Incident['timestamp']): DateIndication {
    if (!value) return ''
    return dateIndicationValue[typeof value]
  }

  const updateTimestamp = useCallback(
    (event) => {
      const { name, value: targetValue } = event.target
      const cloned = new Date(datetime)
      switch (name) {
        case 'day': {
          const dateStr = `${targetValue} ${cloned.getFullYear()}`
          const parsedDate = parse(dateStr, 'EEEE d MMMM yyyy', datetime, {
            locale,
          })
          parsedDate.setHours(cloned.getHours())
          parsedDate.setMinutes(cloned.getMinutes())

          cloned.setTime(parsedDate.getTime())

          break
        }

        case 'hours':
          cloned.setHours(targetValue)
          break
        case 'minutes':
          cloned.setMinutes(targetValue)
          break
      }

      setDatetime(cloned)
      onUpdate(cloned.getTime())
    },
    [datetime, onUpdate]
  )

  const updateIndication = useCallback(
    (indication: DateIndicator['id']) => {
      setDateIndication(indication)

      if (indication === 'now') {
        const cloned = new Date()

        cloned.setHours(9)
        cloned.setMinutes(0)
        cloned.setSeconds(0)

        setDatetime(cloned)
        onUpdate(null)
      } else {
        setDatetime(defaultTimestamp)
        onUpdate(defaultTimestamp.getTime())
      }
    },
    [onUpdate]
  )

  useEffect(() => {
    setDateIndication(getDateIndication(value))
  }, [value])
  return (
    <>
      <RadioGroup>
        {Object.values(dateIndicators).map(({ id, label }) => (
          <Label key={id} label={label} noActiveState>
            <Radio
              checked={id === dateIndication}
              id={id}
              name="dateIndicator"
              onChange={() => updateIndication(id)}
              value={id}
            />
          </Label>
        ))}
      </RadioGroup>

      {dateIndication === 'earlier' && (
        <>
          <TimeFieldset>
            <FieldWrapper>
              <StyledLabel htmlFor="day">Welke dag was het?</StyledLabel>
              <Select
                id="day"
                name="day"
                data-testid="selectDay"
                value={getFormattedDate(datetime)}
                onChange={updateTimestamp}
                options={daysOptions}
              />
            </FieldWrapper>
          </TimeFieldset>

          <TimeFieldset>
            <legend>
              <StyledLabel>Hoe laat was het?</StyledLabel>
            </legend>

            <TimeWrapper>
              <div>
                <Select
                  id="hours"
                  aria-labelledby="uur"
                  name="hours"
                  data-testid="selectHours"
                  value={datetime.getHours().toString()}
                  onChange={updateTimestamp}
                  options={hoursOptions}
                />
              </div>
              <Info id="uur">uur</Info>
              <div>
                <Select
                  id="minutes"
                  name="minutes"
                  aria-labelledby="min"
                  data-testid="selectMinutes"
                  value={datetime.getMinutes().toString()}
                  onChange={updateTimestamp}
                  options={minutesOptions}
                />
              </div>
              <Info id="min" aria-label="minuten">
                min
              </Info>
            </TimeWrapper>
          </TimeFieldset>
        </>
      )}
    </>
  )
}

export default DateTime
