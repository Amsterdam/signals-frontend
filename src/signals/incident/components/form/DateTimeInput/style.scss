.datetime-input {
  &__earlier-date {
    width: 240px;
  }

  &__earlier-time {
    margin-top: 8px;
    display: flex;
    align-items: flex-end;

    select {
      width: 80px !important;
    }

    &-label {
      margin: 0 16px 0 8px;
    }
  }
}
