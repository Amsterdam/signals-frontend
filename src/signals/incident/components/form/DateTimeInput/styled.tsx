import { themeSpacing } from '@amsterdam/asc-ui'
import styled from 'styled-components'

import Label from 'components/Label'

export const StyledLabel = styled(Label)`
  margin-bottom: 0;
  line-height: ${themeSpacing(6)};
`

export const Info = styled.span`
  margin: ${themeSpacing(0, 4, 0, 2)};
`

export const FieldWrapper = styled.div`
  width: 240px; /* fixed value from design */

  & > div:last-child {
    margin-top: ${themeSpacing(3)};
  }
`

export const TimeFieldset = styled.fieldset`
  border: 0;
  padding: 0;
  margin-top: ${themeSpacing(7)};

  & > legend {
    border: 0;
    padding: 0;
  }
`

export const TimeWrapper = styled.div`
  display: flex;
  align-items: flex-end;

  & > div:first-child > div:last-child {
    margin-top: ${themeSpacing(3)};
  }

  select {
    width: 75px; /* fixed value from design */
  }
`
