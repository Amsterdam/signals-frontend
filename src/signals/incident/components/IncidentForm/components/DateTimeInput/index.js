import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';

import Header from '../Header/';

import './style.scss';

const DateTimeInput = ({ touched, hasError, meta, parent, getError, validatorsOrOpts }) => {
  function formatDate(offset, type = 'value') {
    const format = type === 'label' ? 'dddd D MMMM' : 'YYYY-MM-DD';
    if (offset === 0) {
      return 'Vandaag';
    }
    return moment().subtract(offset, 'days').format(format);
  }

  return (
    <div className={`${meta && meta.isVisible ? 'row' : ''}`}>
      {meta && meta.isVisible ?
        <div className={`${meta.className || 'col-12'} mode_input datetime-input`}>
          <div className="invoer antwoorden">
            <Header
              meta={meta}
              options={validatorsOrOpts}
              touched={touched}
              hasError={hasError}
              getError={getError}
            >
              <div className="datetime-input__earlier">
                <div className="datetime-input__label">
                  <label htmlFor={`${meta.name}-select-day`}>Dag</label>
                </div>
                <div className="invoer datetime-input__earlier-date">
                  <select
                    id={`${meta.name}-select-day`}
                    className="datetime-input__earlier-day"
                    value={parent.value.incident_date}
                    onChange={(e) => meta.updateIncident && parent.meta.setIncident({ incident_date: e.target.value })}
                  >
                    {[...Array(7).keys()].map((offset) => (
                      <option key={`select-day-option-${offset}`} value={formatDate(offset)}>{formatDate(offset, 'label')}</option>
                    ))}
                  </select>
                </div>

                <div className="datetime-input__label">
                  <label htmlFor={`${meta.name}-select-time-hours`}>Tijd</label>
                </div>
                <div className="invoer datetime-input__earlier-time">
                  <select
                    value={parent.value.incident_time_hours}
                    id={`${meta.name}-select-time-hours`}
                    className="datetime-input__earlier-time-hours"
                    onChange={(e) => meta.updateIncident && parent.meta.setIncident({ incident_time_hours: e.target.value })}
                  >
                    {[...Array(24).keys()].map((hour) => (
                      <option
                        key={`select-time-hours-option-${hour}`}
                        value={hour}
                      >{hour}</option>
                    ))}
                  </select>
                  <span className="datetime-input__earlier-time-label">uur</span>
                  <select
                    id={`${meta.name}-select-time-minutes`}
                    className="datetime-input__earlier-time-minutes"
                    value={parent.value.incident_time_minutes}
                    onChange={(e) => meta.updateIncident && parent.meta.setIncident({ incident_time_minutes: e.target.value })}
                  >
                    {[...Array(12).keys()].map((minute) => (
                      <option
                        key={`select-time-minutes-option-${minute * 5}`}
                        value={minute * 5}
                      >{minute * 5}</option>
                    ))}
                  </select>
                  <span className="datetime-input__earlier-time-label">min</span>
                </div>
              </div>
            </Header>
          </div>
        </div>
         : ''}
    </div>
  );
};

DateTimeInput.propTypes = {
  touched: PropTypes.bool,
  hasError: PropTypes.func,
  meta: PropTypes.object,
  parent: PropTypes.object,
  getError: PropTypes.func,
  validatorsOrOpts: PropTypes.object
};

export default DateTimeInput;
