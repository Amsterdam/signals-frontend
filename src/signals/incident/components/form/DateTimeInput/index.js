import React from 'react';
import format from 'date-fns/format';
import subDays from 'date-fns/subDays';
import PropTypes from 'prop-types';
import { nl } from 'date-fns/locale';
import Select from 'components/SelectInput';
import { capitalize } from 'shared/services/date-utils';
import Header from '../Header';

import './style.scss';

const formatDate = (offset, type = 'value') => {
  const dateFormat = type === 'label' ? 'EEEE d MMMM' : 'yyyy-MM-dd';
  if (offset === 0) {
    return 'Vandaag';
  }

  const date = subDays(new Date(), offset);
  return capitalize(format(date, dateFormat, { locale: nl }));
};


const DateTimeInput = ({ touched, hasError, meta, parent, getError, validatorsOrOpts }) => {
  if (!meta?.isVisible) return null;

  const options = [...Array(7).keys()].map(offset => {
    const name = formatDate(offset, 'label');
    const value = formatDate(offset);
    return {
      value,
      key: name,
      name,
    };
  });

  return (
    <Header meta={meta} options={validatorsOrOpts} touched={touched} hasError={hasError} getError={getError}>
      <div className="datetime-input__earlier">
        <div className="datetime-input__earlier-date">
          <Select
            name="day"
            data-testid="selectDay"
            label={<strong>Dag</strong>}
            value={`${parent.value.incident_date}`}
            onChange={e => parent.meta.updateIncident({ incident_date: e.target.value })}
            options={options}
          />
        </div>

        <div className="datetime-input__earlier-time">
          <div>
            <Select
              name="hours"
              data-testid="selectHours"
              label={<strong>Tijd</strong>}
              value={`${parent.value.incident_time_hours}`}
              onChange={e => parent.meta.updateIncident({ incident_time_hours: e.target.value })}
              options={[...Array(24).keys()].map(value => ({ value, key: value, name: value }))}
            />
          </div>
          <span className="datetime-input__earlier-time-label">uur</span>
          <div>
            <Select
              name="minutes"
              data-testid="selectMinutes"
              value={`${parent.value.incident_time_minutes}`}
              onChange={e => parent.meta.updateIncident({ incident_time_minutes: e.target.value })}
              options={[...Array(12).keys()].map(minute => ({
                value: minute * 5,
                name: minute * 5,
                key: minute * 5,
              }))}
            />
          </div>
          <span className="datetime-input__earlier-time-label">min</span>
        </div>
      </div>
    </Header>
  );
};

DateTimeInput.defaultProps = {
  hasError: () => {},
};

DateTimeInput.propTypes = {
  touched: PropTypes.bool,
  hasError: PropTypes.func,
  meta: PropTypes.object,
  parent: PropTypes.object,
  getError: PropTypes.func,
  validatorsOrOpts: PropTypes.object,
};

export default DateTimeInput;
