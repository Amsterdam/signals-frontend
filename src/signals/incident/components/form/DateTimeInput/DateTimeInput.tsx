// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2018 - 2022 Gemeente Amsterdam
import type { FC } from 'react'

import FormField from 'components/FormField'
import type { FormInputProps } from 'types/reactive-form'

import DateTime from './DateTime'

type DateTimeInputProps = Omit<FormInputProps<number | null>, 'handler'>

const DateTimeInput: FC<DateTimeInputProps> = ({
  hasError,
  meta,
  parent,
  getError,
  validatorsOrOpts,
  value,
}) => {
  if (!meta?.isVisible) return null

  const updateTimestamp = (timestamp: number) => {
    parent.meta.updateIncident({ dateTime: timestamp })
  }

  return (
    <FormField
      meta={meta}
      options={validatorsOrOpts}
      hasError={hasError}
      getError={getError}
      isFieldSet
    >
      <DateTime onUpdate={updateTimestamp} value={value} />
    </FormField>
  )
}

export default DateTimeInput
