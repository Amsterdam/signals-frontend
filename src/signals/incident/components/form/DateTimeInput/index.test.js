import React from 'react';
import { shallow } from 'enzyme';

import DateTimeInput from './index';

describe('Form component <DateTimeInput />', () => {
  const metaFields = {
    name: 'input-field-name',
  };
  let wrapper;
  let parent;

  const RealDate = Date;

  const mockDate = isoDate => {
    global.Date = class extends RealDate {
      constructor(...theArgs) {
        if (theArgs.length) {
          return new RealDate(...theArgs);
        }
        return new RealDate(isoDate);
      }

      static now() {
        return new RealDate(isoDate).getTime();
      }
    };
  };

  afterEach(() => {
    global.Date = RealDate;
  });

  beforeEach(() => {
    mockDate('2017-02-14');

    parent = {
      meta: {
        updateIncident: jest.fn(),
      },
      value: {
        incident_time_hours: 9,
        incident_time_minutes: 0,
      },
    };

    wrapper = shallow(<DateTimeInput parent={parent} />);
  });


  describe('rendering', () => {
    it('should render date time field correctly', () => {
      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: true,
        },
      });

      expect(wrapper).toMatchSnapshot();
    });

    it('should render no date time field when not visible', () => {
      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: false,
        },
      });

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('events', () => {
    it('sets incident day when value changes', () => {
      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: true,
        },
      });

      wrapper.find('[data-testid="selectDay"]').simulate('change', { target: { value: '2018-07-21' } });

      expect(parent.meta.updateIncident).toHaveBeenCalledWith({
        incident_date: '2018-07-21',
      });
    });

    it('sets incident hour when value changes', () => {
      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: true,
        },
      });

      wrapper.find('[data-testid="selectHours"]').simulate('change', { target: { value: '13' } });

      expect(parent.meta.updateIncident).toHaveBeenCalledWith({
        incident_time_hours: '13',
      });
    });

    it('sets incident minutes when value changes', () => {
      wrapper.setProps({
        meta: {
          ...metaFields,
          isVisible: true,
        },
      });

      wrapper.find('[data-testid="selectMinutes"]').simulate('change', { target: { value: '42' } });

      expect(parent.meta.updateIncident).toHaveBeenCalledWith({
        incident_time_minutes: '42',
      });
    });
  });
});
