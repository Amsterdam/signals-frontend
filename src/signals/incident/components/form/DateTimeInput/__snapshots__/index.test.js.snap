// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component <DateTimeInput /> rendering should render date time field correctly 1`] = `
<Header
  className=""
  hasError={[Function]}
  meta={
    Object {
      "isVisible": true,
      "name": "input-field-name",
    }
  }
>
  <div
    className="datetime-input__earlier"
  >
    <div
      className="datetime-input__earlier-date"
    >
      <SelectInput
        data-testid="selectDay"
        label={
          <strong>
            Dag
          </strong>
        }
        name="day"
        onChange={[Function]}
        options={
          Array [
            Object {
              "key": "Vandaag",
              "name": "Vandaag",
              "value": "Vandaag",
            },
            Object {
              "key": "Maandag 13 februari",
              "name": "Maandag 13 februari",
              "value": "2017-02-13",
            },
            Object {
              "key": "Zondag 12 februari",
              "name": "Zondag 12 februari",
              "value": "2017-02-12",
            },
            Object {
              "key": "Zaterdag 11 februari",
              "name": "Zaterdag 11 februari",
              "value": "2017-02-11",
            },
            Object {
              "key": "Vrijdag 10 februari",
              "name": "Vrijdag 10 februari",
              "value": "2017-02-10",
            },
            Object {
              "key": "Donderdag 9 februari",
              "name": "Donderdag 9 februari",
              "value": "2017-02-09",
            },
            Object {
              "key": "Woensdag 8 februari",
              "name": "Woensdag 8 februari",
              "value": "2017-02-08",
            },
          ]
        }
        value="undefined"
      />
    </div>
    <div
      className="datetime-input__earlier-time"
    >
      <div>
        <SelectInput
          data-testid="selectHours"
          label={
            <strong>
              Tijd
            </strong>
          }
          name="hours"
          onChange={[Function]}
          options={
            Array [
              Object {
                "key": 0,
                "name": 0,
                "value": 0,
              },
              Object {
                "key": 1,
                "name": 1,
                "value": 1,
              },
              Object {
                "key": 2,
                "name": 2,
                "value": 2,
              },
              Object {
                "key": 3,
                "name": 3,
                "value": 3,
              },
              Object {
                "key": 4,
                "name": 4,
                "value": 4,
              },
              Object {
                "key": 5,
                "name": 5,
                "value": 5,
              },
              Object {
                "key": 6,
                "name": 6,
                "value": 6,
              },
              Object {
                "key": 7,
                "name": 7,
                "value": 7,
              },
              Object {
                "key": 8,
                "name": 8,
                "value": 8,
              },
              Object {
                "key": 9,
                "name": 9,
                "value": 9,
              },
              Object {
                "key": 10,
                "name": 10,
                "value": 10,
              },
              Object {
                "key": 11,
                "name": 11,
                "value": 11,
              },
              Object {
                "key": 12,
                "name": 12,
                "value": 12,
              },
              Object {
                "key": 13,
                "name": 13,
                "value": 13,
              },
              Object {
                "key": 14,
                "name": 14,
                "value": 14,
              },
              Object {
                "key": 15,
                "name": 15,
                "value": 15,
              },
              Object {
                "key": 16,
                "name": 16,
                "value": 16,
              },
              Object {
                "key": 17,
                "name": 17,
                "value": 17,
              },
              Object {
                "key": 18,
                "name": 18,
                "value": 18,
              },
              Object {
                "key": 19,
                "name": 19,
                "value": 19,
              },
              Object {
                "key": 20,
                "name": 20,
                "value": 20,
              },
              Object {
                "key": 21,
                "name": 21,
                "value": 21,
              },
              Object {
                "key": 22,
                "name": 22,
                "value": 22,
              },
              Object {
                "key": 23,
                "name": 23,
                "value": 23,
              },
            ]
          }
          value="9"
        />
      </div>
      <span
        className="datetime-input__earlier-time-label"
      >
        uur
      </span>
      <div>
        <SelectInput
          data-testid="selectMinutes"
          name="minutes"
          onChange={[Function]}
          options={
            Array [
              Object {
                "key": 0,
                "name": 0,
                "value": 0,
              },
              Object {
                "key": 5,
                "name": 5,
                "value": 5,
              },
              Object {
                "key": 10,
                "name": 10,
                "value": 10,
              },
              Object {
                "key": 15,
                "name": 15,
                "value": 15,
              },
              Object {
                "key": 20,
                "name": 20,
                "value": 20,
              },
              Object {
                "key": 25,
                "name": 25,
                "value": 25,
              },
              Object {
                "key": 30,
                "name": 30,
                "value": 30,
              },
              Object {
                "key": 35,
                "name": 35,
                "value": 35,
              },
              Object {
                "key": 40,
                "name": 40,
                "value": 40,
              },
              Object {
                "key": 45,
                "name": 45,
                "value": 45,
              },
              Object {
                "key": 50,
                "name": 50,
                "value": 50,
              },
              Object {
                "key": 55,
                "name": 55,
                "value": 55,
              },
            ]
          }
          value="0"
        />
      </div>
      <span
        className="datetime-input__earlier-time-label"
      >
        min
      </span>
    </div>
  </div>
</Header>
`;

exports[`Form component <DateTimeInput /> rendering should render no date time field when not visible 1`] = `""`;
